### PR TITLE
Vkcommand changes only

### PR DESCRIPTION
**Pull request by phoenixxxx**
As requested splitting the command buffer specific changes into it's own PR.

### CODE REVIEW

File: filament/backend/src/vulkan/VulkanCommands.cpp [hunk 1]
```diff
@@ -33,453 +33,499 @@ using namespace utils;
 
 namespace filament::backend {
 
+    namespace {
```
Comment (poweifeng) [inline filament/backend/src/vulkan/VulkanCommands.cpp:36]: can you move this namespace 4 spaces to the left?

File: filament/backend/src/vulkan/VulkanCommands.cpp [hunk 2]
```diff
@@ -33,453 +33,499 @@ using namespace utils;
...
-    };
-    vkBeginCommandBuffer(currentbuf->buffer(), &binfo);
+        auto const findNext = [this]() {
+            for (int8_t i = 0; i < CAPACITY; ++i) {
+                if (!mSubmitted[i]) {
+                    return i;
+                }
+            }
+            return INVALID;
+            };
```
Comment (poweifeng) [inline filament/backend/src/vulkan/VulkanCommands.cpp:300]: this bracket should line up with "auto"

File: filament/backend/src/vulkan/VulkanCommands.cpp [hunk 3]
```diff
@@ -33,453 +33,499 @@ using namespace utils;
...
-        pushGroupMarker(marker.c_str(), time);
+    void CommandBufferPool::gc() {
+        ActiveBuffers reclaimed;
+        mSubmitted.forEachSetBit([this, &reclaimed](size_t index) {
+            auto& buffer = mBuffers[index];
+            if (buffer->getStatus() == VK_SUCCESS) {
+                reclaimed.set(index, true);
+                buffer->reset();
+            }
+            });
```
Comment (poweifeng) [inline filament/backend/src/vulkan/VulkanCommands.cpp:320]: bracket and parentheses should line up with "mSubmitted"

File: filament/backend/src/vulkan/VulkanCommands.cpp [hunk 4]
```diff
@@ -33,453 +33,499 @@ using namespace utils;
...
-        auto wrapper = mStorage[i].get();
-        if (wrapper->buffer() != VK_NULL_HANDLE
-                && mCurrentCommandBufferIndex != static_cast<int8_t>(i)) {
-            fences[count++] = wrapper->fence->getFence();
+        if (!mProtectedPool) {
+            mProtectedPool = std::make_unique<CommandBufferPool>(mContext, mAllocator, mDevice,
+                mProtectedQueue, mProtectedQueueFamilyIndex, true);
         }
+        auto& ret = mProtectedPool->getRecording();
+        return ret;
```
Comment (poweifeng) [inline filament/backend/src/vulkan/VulkanCommands.cpp:405]: can simplify to `return mProtected->getRecording();`

File: filament/backend/src/vulkan/VulkanCommands.h [hunk 5]
```diff
@@ -37,218 +38,247 @@
 
 namespace filament::backend {
 
-struct VulkanContext;
+    struct VulkanContext;
```
Comment (poweifeng) [inline filament/backend/src/vulkan/VulkanCommands.h:41]: everything should be shifted 4 spaces to the left

File: filament/backend/src/vulkan/VulkanCommands.h [hunk 6]
```diff
@@ -37,218 +38,247 @@
...
-    VkSemaphore mSubmissionSignal = {};
-    VkSemaphore mInjectedSignal = {};
-    utils::FixedCapacityVector<std::unique_ptr<VulkanCommandBuffer>> mStorage;
-    VkFence mFences[CAPACITY] = {};
-    VkSemaphore mSubmissionSignals[CAPACITY] = {};
-    uint8_t mAvailableBufferCount = CAPACITY;
-    CommandBufferObserver* mObserver = nullptr;
+    private:
+
+        static constexpr int CAPACITY = FVK_MAX_COMMAND_BUFFERS;
```
Comment (poweifeng) [inline filament/backend/src/vulkan/VulkanCommands.h:260]: unused

File: filament/backend/src/vulkan/VulkanCommands.h [hunk 7]
```diff
@@ -37,218 +38,247 @@
...
+        VkDevice const mDevice;
+        VkQueue const mProtectedQueue;
+        // For defered initialization if/when we need protected content
```
Comment (poweifeng) [inline filament/backend/src/vulkan/VulkanCommands.h:263]: "deferred"

File: filament/backend/src/vulkan/VulkanCommands.h [hunk 8]
```diff
@@ -37,218 +38,247 @@
...
+        uint32_t const mProtectedQueueFamilyIndex;
+        VulkanResourceAllocator* mAllocator;
+        VulkanContext* mContext;
+
+        // int8 only goes up to 127, therefore capacity must be less than that.
+        static_assert(CAPACITY < 128);
```
Comment (poweifeng) [inline filament/backend/src/vulkan/VulkanCommands.h:269]: Move to CommandBufferPool

File: filament/backend/src/vulkan/VulkanDriver.cpp [hunk 9]
```diff
@@ -199,7 +199,8 @@ VulkanDriver::VulkanDriver(VulkanPlatform* platform, VulkanContext const& contex
       mResourceManager(&mResourceAllocator),
       mThreadSafeResourceManager(&mResourceAllocator),
       mCommands(mPlatform->getDevice(), mPlatform->getGraphicsQueue(),
-              mPlatform->getGraphicsQueueFamilyIndex(), &mContext, &mResourceAllocator),
+              mPlatform->getGraphicsQueueFamilyIndex(),
+          nullptr, 0, &mContext, &mResourceAllocator),
```
Comment (poweifeng) [inline filament/backend/src/vulkan/VulkanDriver.cpp:203]: indentation looks strange.

File: filament/backend/src/vulkan/VulkanDriver.cpp [hunk 10]
```diff
@@ -690,8 +691,14 @@ void VulkanDriver::destroyRenderTarget(Handle<HwRenderTarget> rth) {
...
 void VulkanDriver::createFenceR(Handle<HwFence> fh, int) {
-    VulkanCommandBuffer const& commandBuffer = mCommands.get();
-    mResourceAllocator.construct<VulkanFence>(fh, commandBuffer.fence);
+    VulkanCommandBuffer* cmdbuf = nullptr;
+    if (mCurrentRenderPass.cmdBuffer) {
+        cmdbuf = mCurrentRenderPass.cmdBuffer;
+    }
+    else {
+        cmdbuf = &mCommands.get();
+    }
```
Comment (poweifeng) [inline filament/backend/src/vulkan/VulkanDriver.cpp:700]: let's not add this yet.

File: filament/backend/src/vulkan/VulkanContext.h [hunk 11]
```diff
@@ -82,6 +82,7 @@ class VulkanTimestamps {
 };
 
 struct VulkanRenderPass {
+    VulkanCommandBuffer* cmdBuffer;
```
Comment (poweifeng) [inline filament/backend/src/vulkan/VulkanContext.h:85]: let's move this to another PR.

**Comment by phoenixxxx** [comment]
Addressed
